### DISCUSSION TITLE
How can I disable all instrumentation related to metrics and logs in OpenTelemetry Java Agent, enabling only traces?

### DISCUSSION THREAD

**Discussion by shahparthiv** [open-telemetry/opentelemetry-java-instrumentation#12537] [Q&A]
I'm using the OpenTelemetry Java Agent to instrument my application, but I only want to instrument traces. Currently, the agent also instruments logs and metrics, which I’d like to disable to reduce overhead and focus purely on tracing.

Could someone guide me on how to configure the OpenTelemetry Java Agent so that:

1. Metrics instrumentation is completely disabled and no metrics data is exported.
2. Logging instrumentation is disabled, so no logs are automatically captured or emitted by the agent.

In short, I want the agent to only handle tracing without any additional instrumentation for logs and metrics.

I’ve tried setting a few properties but am unsure if I’m missing anything or if there’s an all-encompassing way to achieve this. Any guidance or recommended configuration settings would be much appreciated!

**Comment by shahparthiv** [comment]
@srinusanduri , according to the following link, I believe you have worked on similar stuff.  As mentioned in the comment attached in the following link you have achieved to disable metrics and logs and only enabled the traces. Your help would be really appreciated!

https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/10284#discussioncomment-8362573

**Comment by shahparthiv** [comment]
@jack-berg , will this https://github.com/open-telemetry/opentelemetry-java/pull/6549 is useful here?

**Comment by trask** [comment]
did you want more than `OTEL_METRICS_EXPORTER=none OTEL_LOGS_EXPORTER=none`?

**Reply by shahparthiv** [reply]
Thanks for your quick response. I agree but still the agent will instrument it but will not export it. I don’t want to instrument it at all!! any chance??

**Reply by trask** [reply]
this lays some of the groundwork needed for disabling metrics: https://github.com/open-telemetry/opentelemetry-java/pull/6502

for disabling log instrumentation check out https://opentelemetry.io/docs/zero-code/java/agent/disable/#suppressing-specific-agent-instrumentation

**Reply by shahparthiv** [reply]
It is interesting. As I am new to java otel agent could you please help me with a couple of VM properties to disable metrics instrumentation using https://github.com/open-telemetry/opentelemetry-java/pull/6502 or some examples? It would be a great help! 

Thanks in advance.

**Reply by jack-berg** [reply]
> I agree but still the agent will instrument it but will not export it. I don’t want to instrument it at all!! any chance??

Will defer to the @open-telemetry/java-instrumentation-maintainers for the final say, but this isn't really possible today. As of now, the java agent's metric and trace instrumentation are pretty tightly intertwined, such that if a particular instrumentation library is enabled, its going to call the APIs for both traces and metrics.

As @trask notes, https://github.com/open-telemetry/opentelemetry-java/pull/6502 lays the groundwork for a more nuanced approach here. The new APIs to check if a tracer or metric instrument are enabled could be called by instrumentation libraries, and skip recording altogether if disabled. But the instrumentation doesn't do this today. 

For now, instrumentation libraries always record both trace and metrics if a particular library is enabled. However, if there is no exporter configured for a given signal, (i.e. `OTEL_METRICS_EXPORTER=none`), then the SDK will substitute a low-overhead noop implementation. The calls to record the telemetry are still made, but their impact is minimized.

**Reply by shahparthiv** [reply]
@jack-berg , thanks for your reply,

As of now, I have tried to debug the OTEL java agent and my findings are as below, please correct me if I am missing anything. You are 100% correct in the case of logs exporter as I have seen the `NoopLogRecordProcessor` while debugging but,

I found the following whenever I set the metrics exporter as none by setting up the `OTEL_METRICS_EXPORTER=none`.

<img width="1299" alt="Screenshot 2024-11-04 at 11 59 58 AM" src="https://github.com/user-attachments/assets/0182648d-793b-455c-b2bb-2101b0857a78">

Here you can see the metricReaders and metricProducers, lists are empty in metricProvider and according to  https://opentelemetry.io/docs/specs/otel/metrics/api/, the metricProvider is the entrypoint for metric instrumentation. 

So can we confirm that the application is not instrumenting whenever we have the metrics exporter as none?
Your input here will be appreciated.
